shell/platform/android/io/flutter/embedding/engine/mutatorsstack/FlutterMutatorView.java

```diff
@@ -0,0 +1,89 @@
+import android.graphics.Matrix;
+import android.graphics.Path;
+import android.widget.FrameLayout;
+import androidx.annotation.NonNull;
+
+public class FlutterMutatorView extends FrameLayout {
+
+  private FlutterMutatorsStack mutatorsStack;
+  private float screenDensity;
+  private int left;
+  private int top;
+
+  public FlutterMutatorView(@NonNull Context context) {
+    super(context, null);
+  }
+
+  public void readyToDisplay(
+      @NonNull FlutterMutatorsStack mutatorsStack,
+      float screenDensity,
+      int left,
+      int top,
+      int width,
+      int height) {
+    this.mutatorsStack = mutatorsStack;
+    this.screenDensity = screenDensity;
+    this.left = left;
+    this.top = top;
+    FrameLayout.LayoutParams layoutParams = new FrameLayout.LayoutParams(width, height);
+    layoutParams.leftMargin = left;
+    layoutParams.topMargin = top;
+    setLayoutParams(layoutParams);
+    setClipChildren(false);
+    setWillNotDraw(false);
+  }
+
+  @Override
+  public void draw(Canvas canvas) {
+    // Apply all clippings on the parent canvas.
+    canvas.save();
+    for (Path path : mutatorsStack.getFinalClippingPaths()) {
+      // Reverse the current offset.
+      //
+      // The frame of this view includes the final offset of the bounding rect.
+      // We need to apply all the mutators to the view, which includes the mutation that leads to
+      // the final offset. We should reverse this final offset, both as a translate mutation and to
+      // all the clipping paths
+      Path pathCopy = new Path(path);
+      pathCopy.offset(-left, -top);
+      canvas.clipPath(pathCopy);
+    }
+    super.draw(canvas);
+    canvas.restore();
+  }
+
+  @Override
+  public void dispatchDraw(Canvas canvas) {
+    // Apply all the transforms on the child canvas.
+    canvas.save();
+    // Reverse the current offset.
+    //
+    // The frame of this view includes the final offset of the bounding rect.
+    // We need to apply all the mutators to the view, which includes the mutation that leads to
+    // the final offset. We should reverse this final offset, both as a translate mutation and to
+    // all the clipping paths
+    Matrix reverseTranslateMatrix = new Matrix();
+    reverseTranslateMatrix.preTranslate(-left, -top);
+
+    // Reverse scale based on screen scale.
+    //
+    // The Android frame is set based on the logical resolution instead of physical.
+    // (https://developer.android.com/training/multiscreen/screendensities).
+    // However, flow is based on the physical resolution. For example, 1000 pixels in flow equals
+    // 500 points in Android. And until this point, we did all the calculation based on the flow
+    // resolution. So we need to scale down to match Android's logical resolution.
+    Matrix reverseScaleMatrix = new Matrix();
+    reverseScaleMatrix.preScale(1 / screenDensity, 1 / screenDensity);
+
+    Matrix finalMatrix = new Matrix(mutatorsStack.getFinalMatrix());
```

Comment:
what about just allocating a single `Matrix`, and then calling `preScale`  and `preTranslate`?

Reply:
done